packages/fetchai/connections/prometheus/connection.py

```diff
@@ -0,0 +1,341 @@
+
+    async def send(self, envelope: Envelope) -> None:
+        """
+        Process the envelopes to prometheus.
+
+        :return: None
+        """
+        sender = envelope.sender
+        self.logger.debug("Processing message from {}: {}".format(sender, envelope))
+        if envelope.protocol_id != PrometheusMessage.protocol_id:
+            raise ValueError("This protocol is not valid for prometheus.")
+        await self._handle_prometheus_message(envelope)
+
+    async def _handle_prometheus_message(self, envelope: Envelope) -> None:
+        """
+        Handle messages to prometheus.
+
+        :param envelope: the envelope
+        :return: None
+        """
+        enforce(
+            isinstance(envelope.message, PrometheusMessage),
+            "Message not of type PrometheusMessage",
+        )
+        message, dialogue = self._get_message_and_dialogue(envelope)
+
+        if dialogue is None:
+            self.logger.warning(
+                "Could not create dialogue from message={}".format(message)
+            )
+            return
+
+        if message.performative == PrometheusMessage.Performative.ADD_METRIC:
+            response = await self._handle_add_metric(message)
+        elif message.performative == PrometheusMessage.Performative.UPDATE_METRIC:
+            response = await self._handle_update_metric(message)
+        else:  # pragma: nocover
+            self.logger.warning("Unrecognized performative for PrometheusMessage")
+            return
+
+        response_code, response_msg = cast(Tuple[int, str], response)
+
+        msg = dialogue.reply(
+            performative=PrometheusMessage.Performative.RESPONSE,
+            target_message=message,
+            code=response_code,
+            message=response_msg,
+        )
+        context = cast(EnvelopeContext, envelope.context)
+        envelope = Envelope(
+            to=msg.to,
+            sender=msg.sender,
+            protocol_id=msg.protocol_id,
+            message=msg,
+            context=context,
+        )
+        await self._send(envelope)
+
+    async def _handle_add_metric(self, message: PrometheusMessage) -> Tuple[int, str]:
+        """Handle add metric message.
+
+        :param message: the message to handle.
+        :return: the response code and response message.
+        """
+        if message.title in self.metrics:
+            response_code = 409
+            response_msg = "Metric already exists."
+        else:
+            metric_type = getattr(aioprometheus, message.type, None)
+            if metric_type is None:
+                response_code = 404
+                response_msg = f"{message.type} is not a recognized prometheus metric."
+            else:
+                self.metrics[message.title] = metric_type(
+                    message.title, message.description, message.labels
+                )
+                self._service.register(self.metrics[message.title])
```

Comment:
is there no async equivalent?

Reply:
Can this fail?

Reply:
Added a check here to make sure this is a valid metric_type and not some other attribute.

According to the example here (https://github.com/claws/aioprometheus), I believe this is correct usage is async context. But maybe @solarw can confirm?